From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mason Freed <masonf@chromium.org>
Date: Fri, 7 Aug 2026 17:58:25 -0700
Subject: [M152] Add defensive copy in CollectFlattenedAssignedNodes

Original change's description:
> Add defensive copy in CollectFlattenedAssignedNodes
>
> CollectFlattenedAssignedNodes binds a reference to assigned_nodes_ and
> iterates over it by reference while synchronously recursing into slot
> assignment recalculation. A nested recalculation can clear assigned
> nodes of the outer slot, freeing the backing vector during iteration.
>
> This change creates a defensive copy of assigned_nodes to ensure the
> vector backing remains valid throughout the loop iteration.
>
> This fix was validated using the POC provided in the bug description,
> which was implemented and executed locally, but no committable repro
> was found.
>
> Fixed: 540100588
> Change-Id: I39d6e02c7face470cbfe42c6419c5f2ec15519ae
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8175731
> Reviewed-by: Joey Arhar <jarhar@chromium.org>
> Commit-Queue: Mason Freed <masonf@chromium.org>
> Auto-Submit: Mason Freed <masonf@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1672756}

(cherry picked from commit 4de277c7202e0f7e9a2999decffcb0344b3e883d)

Bug: 542457243,540100588
Change-Id: I39d6e02c7face470cbfe42c6419c5f2ec15519ae
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8223369
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Auto-Submit: chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Commit-Queue: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Cr-Commit-Position: refs/branch-heads/7977@{#938}
Cr-Branched-From: b7fe14017379ddffae396d944fb8b59a5896c261-refs/heads/main@{#1669021}

diff --git a/third_party/blink/renderer/core/html/html_slot_element.cc b/third_party/blink/renderer/core/html/html_slot_element.cc
index 5da4aab5323c1ac0f4eac0a9038f58f443df7fa1..66b0228b36988360b8fa8bccb8b309e51c4e5e49 100644
--- a/third_party/blink/renderer/core/html/html_slot_element.cc
+++ b/third_party/blink/renderer/core/html/html_slot_element.cc
@@ -94,7 +94,8 @@ HeapVector<Member<Node>> CollectFlattenedAssignedNodes(
     const HTMLSlotElement& slot) {
   DCHECK(slot.SupportsAssignment());
 
-  const HeapVector<Member<Node>>& assigned_nodes = slot.AssignedNodes();
+  // Copy the array - code inside the loop can modify assigned nodes.
+  const HeapVector<Member<Node>> assigned_nodes = slot.AssignedNodes();
   HeapVector<Member<Node>> nodes;
   if (assigned_nodes.empty()) {
     // Fallback contents.
